File: sdk/pythonsymbolchain/core/nis1/Network.py
import base64

import sha3

from ..ByteArray import ByteArray
from ..Network import Network as BasicNetwork


class Address(ByteArray):
    """Represents a nis address."""

    SIZE = 25

    def __init__(self, address):
        """Creates an address from a decoded or encoded address."""
        raw_bytes = address
        if isinstance(address, str):
            raw_bytes = base64.b32decode(address)
        elif isinstance(address, Address):
            raw_bytes = address.bytes

        super().__init__(self.SIZE, raw_bytes, Address)

    def __str__(self):
        return base64.b32encode(self.bytes).decode('utf8')


class Network(BasicNetwork):
    """Represents a nis network."""

    def address_hasher(self):
        return sha3.keccak_256()

    def create_address(self, address_without_checksum, checksum):
        return Address(address_without_checksum + checksum)


Network.MAINNET = Network('mainnet', 0x68)
Network.TESTNET = Network('testnet', 0x98)
Network.NETWORKS = [Network.MAINNET, Network.TESTNET]

File: sdk/java/src/main/java/org/symbol/sdk/nem/Network.java
package org.symbol.sdk.nem;

import java.time.Instant;
import java.util.List;

import org.bouncycastle.crypto.digests.KeccakDigest;

import org.symbol.sdk.NetworkTimestamp.NetworkTimestampDatetimeConverter;
import org.symbol.sdk.NetworkTimestamp.NetworkTimestampDatetimeConverter.TimeUnit;
import org.symbol.sdk.utils.ArrayHelpers;

/**
 * Represents a NEM network.
 */
public final class Network extends org.symbol.sdk.Network<Address, NetworkTimestamp> {
	private static final Instant NEM_EPOCH = Instant.parse("2015-03-29T00:06:25Z");

	/** NEM main network. */
	public static final Network MAINNET = new Network("mainnet", (byte) 0x68, NEM_EPOCH);

	/** NEM test network. */
	public static final Network TESTNET = new Network("testnet", (byte) 0x98, NEM_EPOCH);

	/** NEM well-known networks. */
	public static final List<Network> NETWORKS = List.of(MAINNET, TESTNET);

	/**
	 * Creates a new NEM network.
	 *
	 * @param name Network name.
	 * @param identifier Network identifier byte.
	 * @param epochTime Network epoch time.
	 */
	public Network(final String name, final byte identifier, final Instant epochTime) {
		super(name, identifier, new NetworkTimestampDatetimeConverter(epochTime, TimeUnit.SECONDS), () -> new KeccakDigest(256),
				// NEM addresses use the full 4-byte checksum the base derives
				(addressWithoutChecksum, checksum) -> new Address(ArrayHelpers.concat(addressWithoutChecksum, checksum)),
				Address.ENCODED_SIZE, Address::new, timestamp -> new NetworkTimestamp((long) timestamp));
	}
}
